river/tree/mondrian/mondrian_tree_nodes.py

```diff
@@ -0,0 +1,474 @@
+import typing
+
+from abc import ABC, abstractmethod
+
+from river.utils.math import log_sum_2_exp
+
+from math import log
+
+from river.tree.base import Leaf
+from river.tree.base import Branch
+
+
+class MondrianTreeLeaf(Leaf, ABC):
+    """
+    Abstract class for all types of nodes in a Mondrian Tree
+
+    Parameters
+    ----------
+    parent
+        Parent Node
+    n_features
+        Number of features of the input data
+    time
+        Split time of the node for Mondrian process
+    """
+
+    def __init__(
+            self,
+            parent: Leaf or None,
+            n_features: int,
+            time: float
+    ):
+
+        super().__init__()
+
+        # Generic Node attributes
+        self.parent = parent
+        self.time = time
+        self.is_leaf = True
+        self.depth = 0
+        self._left = None
+        self._right = None
+        self.feature = 0
+        self.weight = 0.0
+        self.log_weight_tree = 0
+        self.threshold = 0.0
+        self.n_samples = 0
+        self.n_features = n_features
+        self.memory_range_min = [0] * n_features
+        self.memory_range_max = [0] * n_features
+
+    def copy(self, node):
```

Comment:
Can you elaborate on the usage of this one? Maybe `self.clone` can do the trick 🤔

Reply:
This function actually serves to copy characteristics of the node that are not related to samples into a child node when splitting. I could probably use `self.clone` and reset the characteristics that are related to samples the other way around (like instead of saying `child.copy(parent)` I would say `child = parent.clone(); child.n_samples = 0; child.memory_range_...`). Please tell me what would suit you best =)

For my part, I like the logic of having that "copy" function here because it's used for the Regressor as well, and it really is how I want nodes to be copied into another when passing the information from one node to the other, and it would avoid repeating the code for the regressor..

Reply:
Actually, the clone method creates a fresh copy of the object. If you set `include_parameters=True` then the property values will also be copied. Let's discuss this design decision once I got the gist of the code as a whole.